### PR TITLE
Resume state

### PR DESCRIPTION
**Pull request by tombena**
When using option -i followed by filename,
each rank resumes work on the given file
and restores metadata state from a series
of files named "{file}-md-{rank}".

This is particularly useful when the run
starts with a read.

Note: use option -k to keep file during the
previous run

### CODE REVIEW

File: yapio.c [hunk 1]
```diff
@@ -942,6 +945,15 @@ yapio_getopts(int argc, char **argv)
         case 'h':
             yapio_print_help(YAPIO_EXIT_OK);
             break;
+        case 'i':
+            if (optarg[0] == '-')
+            {
+                printf("File not specified with option -i\n");
+                yapio_print_help(YAPIO_EXIT_ERR);
+            }
+            snprintf(yapioTestFileName, PATH_MAX + 1, "%s", optarg);
```
Comment (Nicolas-Triquenaux) [inline yapio.c]: i would check snprintf return as you might truncate the string. Check for -1, and ask for a filename shorter than PATH_MAX

File: yapio.c [hunk 2]
```diff
@@ -1102,15 +1114,20 @@ yapio_open_fpp_file(int rank, int oflags)
...
 yapio_setup_test_file(const yapio_test_group_t *ytg)
 {
+    int path_len;
     yapio_verify_test_directory();
 
-    int path_len = snprintf(yapioTestFileName, PATH_MAX, "%s/%s%d_%s",
+    if (yapioInitFromMdFile)
+        path_len = strlen(yapioTestFileName);
+    else
+        path_len = snprintf(yapioTestFileName, PATH_MAX, "%s/%s%d_%s",
```
Comment (Nicolas-Triquenaux) [inline yapio.c]: i know it is legacy code but later you should check for -1 if the output is truncated

File: yapio.c [hunk 3]
```diff
@@ -1251,9 +1268,36 @@ yapio_alloc_buffers(const yapio_test_group_t *ytg)
         yapio_destroy_buffers_and_abort();
 }
 
+/* read input md file to initialize yapioMyRank's md array */
+static void
+yapio_initialize_source_md_buffer_from_file(const yapio_test_group_t *ytg)
+{
+    /* each rank reads its own md file '{filename}-md-{rank}' */
+    char md_file[PATH_MAX + 10];
```
Comment (Nicolas-Triquenaux) [inline yapio.c]: i hope you never go further than 6 digit numbers (that would be awsome to reach that limit XD)

File: yapio.c [hunk 4]
```diff
@@ -1251,9 +1268,36 @@ yapio_alloc_buffers(const yapio_test_group_t *ytg)
...
+
+    FILE *file = fopen(md_file, "r");
+
+    if (file)
+    {
+        fread(yapioSourceBlkMd, sizeof(yapio_blk_md_t), nblks_per_rank, file);
+        fclose(file);
+    }
+    else
+        printf("Unable to read initial metadata state\n");
```
Comment (Nicolas-Triquenaux) [inline yapio.c]: no "yapio_print_help" to exit in error ?

File: yapio.c [hunk 5]
```diff
@@ -1251,9 +1268,36 @@ yapio_alloc_buffers(const yapio_test_group_t *ytg)
...
+}
+
 static void
 yapio_initialize_source_md_buffer(const yapio_test_group_t *ytg)
 {
+    if (yapioInitFromMdFile)
+    {
+        yapio_initialize_source_md_buffer_from_file(ytg);
+        return;
```
Comment (Nicolas-Triquenaux) [inline yapio.c:1323]: you should not be able to continue if you cannot init the metadata even more if you start with the read operation

File: yapio.c [hunk 6]
```diff
@@ -2135,6 +2180,37 @@ yapio_display_result(const yapio_test_ctx_t *ytc)
     }
 }
 
+/* dump yapioMyRank's metadata array into a file */
+static void
+yapio_store_md_final_state()
+{
+    /* each rank writes its own md file '{filename}-md-{rank}' */
+    char md_file[PATH_MAX + 10];
+    snprintf(md_file, PATH_MAX, "%s-md-%d", yapioTestFileName, yapioMyRank);
```
Comment (Nicolas-Triquenaux) [inline yapio.c]: you should check for the return to see if you truncate the name

File: yapio.c [hunk 7]
```diff
@@ -1251,9 +1268,36 @@ yapio_alloc_buffers(const yapio_test_group_t *ytg)
         yapio_destroy_buffers_and_abort();
 }
 
+/* read input md file to initialize yapioMyRank's md array */
+static void
+yapio_initialize_source_md_buffer_from_file(const yapio_test_group_t *ytg)
```
Comment (Nicolas-Triquenaux) [inline yapio.c:1286]: you do not check that you have the completeness of md file accross all processes. What happen if one of them does not have its md file to read ?

I would add a collective opeation that would sum the number of process that have read successfully its md file.
i would see somewhere an MPI_Allreduce (..... MPI_SUM ....) somewhere, and each process would put 1 in the buffer.

File: yapio.c [hunk 8]
```diff
@@ -2135,6 +2180,37 @@ yapio_display_result(const yapio_test_ctx_t *ytc)
...
+    snprintf(md_file, PATH_MAX, "%s-md-%d", yapioTestFileName, yapioMyRank);
+
+    yapio_test_group_t *ytg = yapioMyTestGroup;
+    const size_t nblks_per_rank = ytg->ytg_num_blks_per_rank;
+
+    /* use metadata from last iteration */
+    int test_ctx_idf = ytg->ytg_num_contexts - 1;
+    yapio_test_ctx_t *ytc = &ytg->ytg_contexts[test_ctx_idf];
+
+    FILE *file = fopen(md_file, "w+");
```
Comment (Nicolas-Triquenaux) [inline yapio.c]: "w+" is really necessary ? just "w" should be fine

File: yapio.c [hunk 9]
```diff
@@ -2135,6 +2180,37 @@ yapio_display_result(const yapio_test_ctx_t *ytc)
...
+
+        const yapio_blk_md_t *md_array =
+            yapio_test_ctx_to_md_array(ytc, YAPIO_TEST_CTX_MDH_OUT, &num_ops_in);
+
+        fwrite(md_array, sizeof(yapio_blk_md_t), nblks_per_rank, file);
+        fclose(file);
+    }
+    else
+        printf("Unable to open file\n");
+}
```
Comment (Nicolas-Triquenaux) [inline yapio.c]: no "yapio_print_help" to exit in error ?

**Review by Nicolas-Triquenaux — CHANGES_REQUESTED** [review]

File: yapio.c (hunk 3 above)
Comment (tombena) [reply yapio.c]: lowering to 4 digit numbers

File: yapio.c (hunk 7 above)
Comment (tombena) [reply yapio.c:1286]: adding it to the bottom of the function

File: yapio.c (hunk 1 above)
Comment (tombena) [reply yapio.c]: done

File: yapio.c (hunk 2 above)
Comment (tombena) [reply yapio.c]: done

File: yapio.c (hunk 4 above)
Comment (tombena) [reply yapio.c]: replacing by fatal log_msg()

File: yapio.c (hunk 5 above)
Comment (tombena) [reply yapio.c:1323]: the program would now crash in yapio_initialize_source_md_buffer_from_file() in this case

File: yapio.c (hunk 6 above)
Comment (tombena) [reply yapio.c]: adding truncate check

File: yapio.c (hunk 8 above)
Comment (tombena) [reply yapio.c]: opening for writing only so "w" should be enough

File: yapio.c (hunk 9 above)
Comment (tombena) [reply yapio.c]: adding it

File: yapio.c [hunk 10]
```diff
@@ -1251,9 +1281,44 @@ yapio_alloc_buffers(const yapio_test_group_t *ytg)
...
+
+    if (file)
+    {
+        n = 1;
+        fread(yapioSourceBlkMd, sizeof(yapio_blk_md_t), nblks_per_rank, file);
+        fclose(file);
+    }
+
+    /* make sure each rank can read its md file */
+    MPI_Allreduce(&n, &global_sum, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
```
Comment (Nicolas-Triquenaux) [inline yapio.c]: hum my bad i said something not optimal. use simple MPI_Reduce like that ; 
http://mpitutorial.com/tutorials/mpi-reduce-and-allreduce/

and have rank0 check the sum. (the sum vs nb of process)

**Review by Nicolas-Triquenaux — CHANGES_REQUESTED** [review]

**Review by Nicolas-Triquenaux — APPROVED** [review]

**Pull request merged**